### PR TITLE
Add retry logic to the produce and consume services

### PR DESCRIPTION
**Pull request by Lincong**

### CODE REVIEW

File: src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2020 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java:2]: Nit: -> `2021`

File: src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.kafka.common.KafkaFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class AbstractService implements Service {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractService.class);
+
+
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java:30]: Nit: Redundant new lines

File: src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    Exception exception = null;
+
+    while (attemptCount < _describeTopicRetries) {
+      DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(Collections.singleton(topic));
+      Map<String, KafkaFuture<TopicDescription>> topicResultValues = describeTopicsResult.values();
+      KafkaFuture<TopicDescription> topicDescriptionKafkaFuture = topicResultValues.get(topic);
+      topicDescription = null;
+      exception = null;
+      try {
+        topicDescription = topicDescriptionKafkaFuture.get();
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/AbstractService.java:58]: Missing return in case successful?

**Review by efeg — CHANGES_REQUESTED** [review]

File: src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java [hunk 4]
```diff
@@ -231,21 +230,13 @@ public synchronized void start() {
...
-        LOG.error("Exception occurred while getting the topicDescriptionKafkaFuture for topic: {}", _topic, e);
-      }
+      TopicDescription topicDescription = getTopicDescription(_adminClient, _topic);
       @SuppressWarnings("ConstantConditions")
       double partitionCount = topicDescription.partitions().size();
-      topicPartitionCount.add(
+      metrics.sensor("topic-partitions").add(
           new MetricName("topic-partitions-count", METRIC_GROUP_NAME, "The total number of partitions for the topic.",
               tags), new CumulativeSum(partitionCount));
+      LOG.info("{}/ConsumeService started", _name);
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java:239]: Should we drop this?

**Review by efeg — CHANGES_REQUESTED** [review]

File: src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java (hunk 4 above)
Comment (Lincong) [reply src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java:239]: Dropped

File: src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java [hunk 5]
```diff
@@ -231,21 +230,13 @@ public synchronized void start() {
...
-      DescribeTopicsResult describeTopicsResult = _adminClient.describeTopics(Collections.singleton(_topic));
-      Map<String, KafkaFuture<TopicDescription>> topicResultValues = describeTopicsResult.values();
-      KafkaFuture<TopicDescription> topicDescriptionKafkaFuture = topicResultValues.get(_topic);
-      TopicDescription topicDescription = null;
-      try {
-        topicDescription = topicDescriptionKafkaFuture.get();
-      } catch (InterruptedException | ExecutionException e) {
-        LOG.error("Exception occurred while getting the topicDescriptionKafkaFuture for topic: {}", _topic, e);
-      }
+      TopicDescription topicDescription = getTopicDescription(_adminClient, _topic);
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/ConsumeService.java:233]: Should we move `_adminClient` and `_topic` to AbstractService and pass no parameter to this method? (similar for ProduceService)

**Review by efeg — CHANGES_REQUESTED** [review]

File: src/main/java/com/linkedin/xinfra/monitor/services/ProduceService.java [hunk 6]
```diff
@@ -156,16 +157,11 @@ private void initializeProducer(Map<String, Object> props) throws Exception {
...
-        _handleNewPartitionsExecutor.scheduleWithFixedDelay(new NewPartitionHandler(), 1, 30, TimeUnit.SECONDS);
-        LOG.info("{}/ProduceService started", _name);
-      } catch (InterruptedException | UnknownTopicOrPartitionException | ExecutionException e) {
-        LOG.error("Exception occurred while starting produce service for topic: {}", _topic, e);
-      }
+      TopicDescription topicDescription = getTopicDescription(_adminClient, _topic);
+      int partitionNum = topicDescription.partitions().size();
+      initializeStateForPartitions(partitionNum);
+      _handleNewPartitionsExecutor.scheduleWithFixedDelay(new NewPartitionHandler(), 1, 30, TimeUnit.SECONDS);
+      LOG.info("{}/ProduceService started", _name);
```
Comment (efeg) [inline src/main/java/com/linkedin/xinfra/monitor/services/ProduceService.java:164]: Should we drop this?

**Review by efeg — CHANGES_REQUESTED** [review]

**Review by efeg — APPROVED** [review]

**Pull request merged**